File: clients/android/app/src/main/java/eu/allodia/mailcal/WritingStyleRevealPages.kt
// The reveal's first three pages: what was read, a typical reply as a miniature letter, and the
// greetings and sign-offs as bars. Each page starts hidden and plays in once it is on screen; the
// arithmetic behind the pictures is in WritingStyleReveal.kt.
package eu.allodia.mailcal

import androidx.compose.animation.core.animateIntAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.ElevatedCard
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.platform.LocalConfiguration
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.rememberTextMeasurer
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.unit.dp
import java.time.Instant
import java.time.ZoneId
import uniffi.mailcal_bindings.HabitRow
import uniffi.mailcal_bindings.LanguageStyleRow
import uniffi.mailcal_bindings.WritingStyleDetail

private val NUMERAL = TextStyle(fontWeight = FontWeight.SemiBold, fontFeatureSettings = "tnum")

// Step 1: how many messages in how many languages since when, from which account.
@Composable
internal fun RevealReadPage(detail: WritingStyleDetail, address: String?, zone: ZoneId) {
    val ctx = LocalContext.current
    val locale = LocalConfiguration.current.locales[0]
    val shown = rememberShown()
    val codes = detail.languages.map { it.language }
    WizardPage(
        title = L10n.reveal_title(ctx),
        bottom = {
            Text(
                L10n.reveal_pages_intro(ctx),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.wizardEntrance(shown, WizardEntrance.RISE, 1_000),
            )
        },
    ) {
        address?.let {
            Text(
                L10n.reveal_based_on(ctx, address = it),
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        val oldest = detail.row.oldest
        val stats = buildList {
            add(RevealStat(L10n.reveal_stat_messages(ctx), detail.row.messages.toInt()))
            add(RevealStat(L10n.reveal_stat_languages(ctx), codes.size))
            oldest?.let { add(RevealStat(L10n.reveal_stat_since(ctx), 0, revealSinceText(it, Instant.now(), zone, locale))) }
        }
        val spoken = oldest?.let {
            L10n.reveal_stats_a11y(
                ctx,
                count = detail.row.messages.toInt(),
                date = epochDate(it, zone, locale),
                languages = writingStyleLanguages(ctx, codes),
            )
        }
        RevealStats(stats, shown, spoken)
        LanguageChips(codes, shown)
    }
}

// One figure of the first page: a count, or a date already worded.
private data class RevealStat(val label: String, val count: Int, val text: String? = null)

// The counts side by side, each under its label, at the largest of three sizes that fits the width.
@Composable
private fun RevealStats(stats: List<RevealStat>, shown: Boolean, spoken: String?) {
    val measurer = rememberTextMeasurer()
    val density = LocalDensity.current
    val label = MaterialTheme.typography.labelLarge
    val base = MaterialTheme.typography.displayMedium.merge(NUMERAL)
    BoxWithConstraints(
        modifier = Modifier
            .fillMaxWidth()
            .padding(top = 8.dp)
            .then(if (spoken != null) Modifier.clearAndSetSemantics { contentDescription = spoken } else Modifier),
    ) {
        val limit = with(density) { maxWidth.toPx() }
        val gaps = with(density) { (STAT_GAP * 2 * (stats.size - 1)).toPx() }
        val sizes = listOf(1f, 0.78f, 0.6f).map { base.copy(fontSize = base.fontSize * it, lineHeight = base.lineHeight * it) }
        val numeral = sizes.firstOrNull { style ->
            stats.sumOf { stat ->
                maxOf(
                    measurer.measure(stat.label, label).size.width,
                    measurer.measure(stat.text ?: stat.count.toString(), style).size.width,
                )
            } + gaps <= limit
        } ?: sizes.last()
        Row(modifier = Modifier.height(IntrinsicSize.Min)) {
            stats.forEachIndexed { index, stat ->
                if (index > 0) {
                    Box(
                        modifier = Modifier
                            .padding(horizontal = STAT_GAP)
                            .width(1.dp)
                            .fillMaxHeight()
                            .background(MaterialTheme.colorScheme.outlineVariant),
                    )
                }
                Column(modifier = Modifier.wizardEntrance(shown, WizardEntrance.RISE, index * 120)) {
                    Text(stat.label, style = label, color = MaterialTheme.colorScheme.onSurfaceVariant)
                    if (stat.text != null) {
                        Text(stat.text, style = numeral)
                    } else {
                        CountingText(stat.count, numeral, shown, index * 120)
                    }
                }
            }
        }
    }
}

private val STAT_GAP = 16.dp

// A count from nought, in a box already as wide as the final value so nothing beside it moves.
@Composable
private fun CountingText(target: Int, style: TextStyle, shown: Boolean, delayMillis: Int) {
    val value by animateIntAsState(if (shown) target else 0, tween(900, delayMillis, COUNT), label = "count")
    Box {
        Text(target.toString(), style = style, modifier = Modifier.alpha(0f))
        Text(value.toString(), style = style)
    }
}

@OptIn(ExperimentalLayoutApi::class)
@Composable
private fun LanguageChips(codes: List<String>, shown: Boolean) {
    val ctx = LocalContext.current
    val colours = with(MaterialTheme.colorScheme) { listOf(primary, tertiary, secondary, inversePrimary, outline) }
    // The counts above already name the languages.
    FlowRow(
        modifier = Modifier.clearAndSetSemantics {},
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        codes.forEachIndexed { index, code ->
            Row(
                modifier = Modifier
                    .wizardEntrance(shown, WizardEntrance.POP, 760 + 80 * index)
                    .heightIn(min = 32.dp)
                    .background(MaterialTheme.colorScheme.surfaceContainerHigh, CircleShape)
                    .padding(start = 12.dp, end = 14.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Box(modifier = Modifier.size(8.dp).background(colours[index % colours.size], CircleShape))
                Text(L10n.languageName(ctx, code), style = MaterialTheme.typography.bodyMedium)
            }
        }
    }
}

// Step 2: a typical reply as a miniature letter, above its length, shape and punctuation.
@Composable
internal fun RevealLetterPage(style: LanguageStyleRow) {
    val ctx = LocalContext.current
    val shown = rememberShown()
    WizardPage(title = L10n.reveal_step_letter(ctx)) {
        RevealLetter(style, shown)
        Text(
            L10n.reveal_letter_caption(ctx),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        if (style.typicalWords > 0u) {
            val count = style.typicalWords.toInt()
            val sentence = L10n.reveal_length(ctx, count)
            val number = count.toString()
            val at = sentence.indexOf(number)
            val large = MaterialTheme.typography.headlineLarge.merge(NUMERAL).toSpanStyle()
            Text(
                buildAnnotatedString {
                    if (at < 0) {
                        append(sentence)
                    } else {
                        append(sentence.substring(0, at))
                        withStyle(large.copy(color = MaterialTheme.colorScheme.onSurface)) { append(number) }
                        append(sentence.substring(at + number.length))
                    }
                },
                style = MaterialTheme.typography.bodyLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier
                    .padding(top = 8.dp)
                    .wizardEntrance(shown, WizardEntrance.RISE, 120)
                    .clearAndSetSemantics { contentDescription = sentence },
            )
        }
        RevealFact(L10n.reveal_shape(ctx), style.shape, shown, 240)
        RevealFact(L10n.reveal_punctuation(ctx), style.punctuation, shown, 360)
    }
}

@Composable
private fun RevealLetter(style: LanguageStyleRow, shown: Boolean) {
    val paragraphs = revealLetterLines(style.typicalWords, style.typicalParagraphs)
    val starts = paragraphs.runningFold(0) { total, lines -> total + lines.size }
    val serif = MaterialTheme.typography.bodyLarge.copy(fontFamily = FontFamily.Serif)
    val ink = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.14f)
    ElevatedCard(
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.elevatedCardColors(containerColor = MaterialTheme.colorScheme.surfaceContainerLowest),
    ) {
        Column(modifier = Modifier.fillMaxWidth().padding(start = 18.dp, end = 18.dp, top = 16.dp, bottom = 18.dp)) {
            style.greetings.firstOrNull()?.let {
                RevealRunsText(
                    revealRuns(it.text),
                    MaterialTheme.typography.headlineSmall.copy(fontFamily = FontFamily.Serif, fontWeight = FontWeight.Medium),
                    Modifier.padding(bottom = 12.dp).wizardEntrance(shown, WizardEntrance.RISE, 40),
                )
            }
            Column(modifier = Modifier.clearAndSetSemantics {}, verticalArrangement = Arrangement.spacedBy(12.dp)) {
                paragraphs.forEachIndexed { paragraph, widths ->
                    Column(verticalArrangement = Arrangement.spacedBy(7.dp)) {
                        widths.forEachIndexed { line, width ->
                            Spacer(
                                modifier = Modifier
                                    .fillMaxWidth()
                                    .height(7.dp)
                                    .growingBar(shown, width.toFloat(), ink, 3.5.dp, 200 + 60 * (starts[paragraph] + line)),
                            )
                        }
                    }
                }
            }
            val signOff = style.signOffs.firstOrNull()
            if (signOff != null || style.signsAs.isNotEmpty()) {
                Column(
                    modifier = Modifier
                        .padding(top = 22.dp)
                        .wizardEntrance(shown, WizardEntrance.RISE, 280 + 60 * starts.last()),
                ) {
                    signOff?.let { RevealRunsText(revealRuns(it.text), serif) }
                    if (style.signsAs.isNotEmpty()) Text(style.signsAs, style = serif)
                }
            }
        }
    }
}

@Composable
private fun RevealFact(label: String, value: String, shown: Boolean, delayMillis: Int) {
    if (value.isEmpty()) return
    Column(
        modifier = Modifier.semantics(mergeDescendants = true) {}.wizardEntrance(shown, WizardEntrance.RISE, delayMillis),
        verticalArrangement = Arrangement.spacedBy(3.dp),
    ) {
        Text(label, style = MaterialTheme.typography.labelLarge, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Text(value, style = MaterialTheme.typography.bodyLarge)
    }
}

// Step 3: greetings and sign-offs, each over a bar as long as its part of its list.
@Composable
internal fun RevealHabitsPage(style: LanguageStyleRow) {
    val ctx = LocalContext.current
    val shown = rememberShown()
    WizardPage(title = L10n.reveal_step_habits(ctx)) {
        HabitGroup(L10n.reveal_greetings(ctx), style.greetings, shown, 80)
        HabitGroup(L10n.reveal_sign_offs(ctx), style.signOffs, shown, 320)
        if (style.signsAs.isNotEmpty()) {
            Row(
                modifier = Modifier
                    .semantics(mergeDescendants = true) {}
                    .wizardEntrance(shown, WizardEntrance.RISE, 560),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Text(
                    L10n.reveal_signs_as(ctx),
                    style = MaterialTheme.typography.labelLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Text(
                    style.signsAs,
                    style = MaterialTheme.typography.titleLarge.copy(fontFamily = FontFamily.Serif),
                )
            }
        }
    }
}

@Composable
private fun HabitGroup(label: String, habits: List<HabitRow>, shown: Boolean, startMillis: Int) {
    if (habits.isEmpty()) return
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Text(
            label,
            style = MaterialTheme.typography.labelLarge,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.semantics { heading() },
        )
        habits.forEachIndexed { index, habit -> HabitBar(habit, top = index == 0, shown, startMillis + 80 * index) }
    }
}

@Composable
private fun HabitBar(habit: HabitRow, top: Boolean, shown: Boolean, delayMillis: Int) {
    val ctx = LocalContext.current
    val shape = RoundedCornerShape(10.dp)
    val colours = MaterialTheme.colorScheme
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .heightIn(min = 40.dp)
            .background(colours.surfaceContainerLow, shape)
            .growingBar(
                shown,
                habit.relative.toFloat() / 100f,
                if (top) colours.primaryContainer else colours.surfaceContainerHighest,
                10.dp,
                delayMillis,
            )
            .semantics(mergeDescendants = true) {}
            .padding(horizontal = 12.dp, vertical = 6.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        RevealRunsText(
            revealRuns(habit.text),
            MaterialTheme.typography.bodyLarge.copy(fontWeight = FontWeight.Medium),
            Modifier.weight(1f),
        )
        Text(
            revealFrequencyText(ctx, habit.frequency),
            style = MaterialTheme.typography.bodyMedium,
            color = colours.onSurfaceVariant,
            modifier = Modifier.padding(start = 8.dp),
        )
    }
}

// A greeting or sign-off, a placeholder such as `[Name]` drawn as a small pill in the accent.
@OptIn(ExperimentalLayoutApi::class)
@Composable
internal fun RevealRunsText(runs: List<RevealRun>, style: TextStyle, modifier: Modifier = Modifier) {
    if (runs.none { it.placeholder }) {
        Text(runs.joinToString("") { it.text }, style = style, modifier = modifier)
        return
    }
    val accent = MaterialTheme.colorScheme.primary
    FlowRow(modifier = modifier, itemVerticalAlignment = Alignment.CenterVertically) {
        runs.forEach { run ->
            if (run.placeholder) {
                Text(
                    run.text,
                    style = MaterialTheme.typography.labelMedium.copy(fontWeight = FontWeight.SemiBold),
                    color = accent,
                    modifier = Modifier
                        .padding(horizontal = 1.dp)
                        .background(accent.copy(alpha = 0.1f), CircleShape)
                        .border(0.5.dp, accent.copy(alpha = 0.4f), CircleShape)
                        .padding(horizontal = 7.dp, vertical = 1.dp),
                )
            } else {
                Text(run.text, style = style)
            }
        }
    }
}

File: clients/apple/Packages/MailcalKit/Sources/MailcalUI/WritingStyleRevealPages.swift
// The reveal's first three pages: what was read, a typical reply as a miniature letter, and the
// greetings and sign-offs as bars. Each page starts hidden and plays in once it is on screen; the
// arithmetic behind the pictures is in WritingStyleReveal.swift.

import Foundation
import MailcalBindings
import SwiftUI

/// Step 1: how many messages in how many languages since when, from which account.
struct RevealReadPage: View {
    let detail: WritingStyleDetail
    /// The source account's address; `nil` when the style came from another device.
    let address: String?
    let zone: TimeZone

    @State private var shown = false
    @Environment(\.colorScheme) private var scheme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @ScaledMetric(relativeTo: .largeTitle) private var numeral: CGFloat = 46

    private static let languageColours: [Color] = [.accentColor, .indigo, .teal, .orange, .pink, .green]

    var body: some View {
        WizardPage(title: L10n.reveal_title()) {
            if let address {
                Text(L10n.reveal_based_on(address: address))
                    .font(WizardFont.text)
                    .foregroundStyle(.secondary)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.top, -6)
            }
            statsRow
                .padding(.top, 18)
            languageChips
                .padding(.top, 14)
            Spacer(minLength: 16)
            Text(L10n.reveal_pages_intro())
                .font(WizardFont.text)
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)
                .wizardEntrance(shown, .rise, delay: 1)
        }
        .onAppear { shown = true }
    }

    private var codes: [String] { detail.languages.map(\.language) }

    private var since: String? {
        detail.row.oldest.map { revealSinceText($0, now: Date(), zone: zone, locale: L10n.appLocale) }
    }

    /// The row read as one sentence, when there is a date to say it with.
    @ViewBuilder
    private var statsRow: some View {
        let row = ViewThatFits(in: .horizontal) {
            stats(numeral)
            stats(numeral * 0.78)
            stats(numeral * 0.6)
        }
        if let oldest = detail.row.oldest {
            row
                .accessibilityElement(children: .ignore)
                .accessibilityLabel(L10n.reveal_stats_a11y(
                    count: Int(detail.row.messages),
                    date: writingStyleDate(oldest, zone: zone, locale: L10n.appLocale),
                    languages: writingStyleLanguages(codes, locale: L10n.appLocale)
                ))
        } else {
            row.accessibilityElement(children: .combine)
        }
    }

    private func stats(_ size: CGFloat) -> some View {
        let font = Font.system(size: size, weight: .semibold, design: .rounded).monospacedDigit()
        return HStack(alignment: .top, spacing: 0) {
            stat(L10n.reveal_stat_messages(), first: true, delay: 0) {
                counting(Int(detail.row.messages), font: font, delay: 0)
            }
            stat(L10n.reveal_stat_languages(), first: false, delay: 0.12) {
                counting(codes.count, font: font, delay: 0.12)
            }
            if let since {
                stat(L10n.reveal_stat_since(), first: false, delay: 0.24) {
                    Text(since).font(font).tracking(-size * 0.025)
                }
            }
        }
        .fixedSize()
    }

    private func stat(
        _ label: String, first: Bool, delay: Double, @ViewBuilder value: () -> some View
    ) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(label).font(WizardFont.label).foregroundStyle(.secondary)
            value()
        }
        .padding(.leading, first ? 0 : 22)
        .padding(.trailing, 22)
        .overlay(alignment: .leading) {
            if !first {
                Rectangle().fill(WizardPalette(scheme).separator).frame(width: 0.5)
            }
        }
        .wizardEntrance(shown, .rise, delay: delay)
    }

    /// A count from nought, in a box already as wide as the final value so nothing beside it moves.
    private func counting(_ target: Int, font: Font, delay: Double) -> some View {
        Text(String(target))
            .font(font)
            .hidden()
            .overlay(alignment: .leading) {
                CountingNumber(value: shown || reduceMotion ? Double(target) : 0) { Text(String($0)) }
                    .font(font)
                    .animation(reduceMotion ? nil : WizardMotion.count(0.9, delay), value: shown)
            }
    }

    private var languageChips: some View {
        let palette = WizardPalette(scheme)
        return RecipientFlowLayout(spacing: 8) {
            ForEach(Array(codes.enumerated()), id: \.offset) { index, code in
                HStack(spacing: 8) {
                    Circle()
                        .fill(Self.languageColours[index % Self.languageColours.count])
                        .frame(width: 8, height: 8)
                    Text(L10n.languageName(code)).font(WizardFont.text)
                }
                .padding(.leading, 10)
                .padding(.trailing, 12)
                .frame(minHeight: 28)
                .background(palette.fill, in: Capsule())
                .wizardEntrance(shown, .pop, delay: 0.76 + 0.08 * Double(index))
            }
        }
        // The row above already names the languages.
        .accessibilityHidden(true)
    }
}

/// Step 2: a typical reply as a miniature letter, beside its length, shape and punctuation.
struct RevealLetterPage: View {
    let style: LanguageStyleRow
    let underTopBar: Bool

    @State private var shown = false
    @Environment(\.colorScheme) private var scheme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @ScaledMetric(relativeTo: .title) private var greetingSize: CGFloat = 28
    @ScaledMetric(relativeTo: .body) private var signSize: CGFloat = 16
    @ScaledMetric(relativeTo: .title) private var lengthSize: CGFloat = 32

    var body: some View {
        WizardPage(title: L10n.reveal_step_letter(), underTopBar: underTopBar) {
            #if os(macOS)
            HStack(alignment: .top, spacing: 22) {
                letterColumn
                facts.frame(width: 215, alignment: .leading)
            }
            #else
            letterColumn
            facts.padding(.top, 2)
            #endif
        }
        .onAppear { shown = true }
    }

    private var lines: [[Double]] {
        revealLetterLines(words: style.typicalWords, paragraphs: style.typicalParagraphs)
    }

    private var letterColumn: some View {
        VStack(alignment: .leading, spacing: 8) {
            letter
            Text(L10n.reveal_letter_caption())
                .font(WizardFont.caption)
                .foregroundStyle(.secondary)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.horizontal, 2)
        }
    }

    private var letter: some View {
        let palette = WizardPalette(scheme)
        let paragraphs = lines
        let starts = paragraphs.reduce(into: [0]) { $0.append($0[$0.count - 1] + $1.count) }
        let serif = Font.system(size: signSize, design: .serif)
        return VStack(alignment: .leading, spacing: 0) {
            if let greeting = style.greetings.first {
                RevealRunsText(
                    runs: revealRuns(greeting.text),
                    font: .system(size: greetingSize, weight: .medium, design: .serif)
                )
                .padding(.bottom, 12)
                .wizardEntrance(shown, .rise, delay: 0.04)
            }
            VStack(alignment: .leading, spacing: 12) {
                ForEach(Array(paragraphs.enumerated()), id: \.offset) { paragraph, widths in
                    VStack(alignment: .leading, spacing: 7) {
                        ForEach(Array(widths.enumerated()), id: \.offset) { line, width in
                            GrowingBar(fraction: shown || reduceMotion ? width : 0, cornerRadius: 3.5)
                                .fill(palette.ink)
                                .frame(height: 7)
                                .animation(
                                    reduceMotion ? nil : WizardMotion.grow(0.2 + 0.06 * Double(starts[paragraph] + line)),
                                    value: shown
                                )
                        }
                    }
                }
            }
            .accessibilityHidden(true)
            if style.signOffs.first != nil || !style.signsAs.isEmpty {
                VStack(alignment: .leading, spacing: 2) {
                    if let signOff = style.signOffs.first {
                        RevealRunsText(runs: revealRuns(signOff.text), font: serif)
                    }
                    if !style.signsAs.isEmpty {
                        Text(style.signsAs).font(serif)
                    }
                }
                .padding(.top, 22)
                .wizardEntrance(shown, .rise, delay: 0.28 + 0.06 * Double(starts[starts.count - 1]))
            }
        }
        .padding(.top, 16)
        .padding(.horizontal, 18)
        .padding(.bottom, 18)
        .frame(maxWidth: .infinity, alignment: .leading)
        .wizardCard(palette, radius: 12)
        .shadow(color: palette.shadow, radius: 14, y: 10)
        .accessibilityElement(children: .combine)
    }

    private var facts: some View {
        VStack(alignment: .leading, spacing: 16) {
            if style.typicalWords > 0 {
                CountingNumber(value: shown || reduceMotion ? Double(style.typicalWords) : 0) { length($0) }
                    .animation(reduceMotion ? nil : WizardMotion.count(0.7, 0.12), value: shown)
                    .fixedSize(horizontal: false, vertical: true)
                    .accessibilityElement(children: .ignore)
                    .accessibilityLabel(L10n.reveal_length(count: Int(style.typicalWords)))
                    .wizardEntrance(shown, .rise, delay: 0.12)
            }
            fact(L10n.reveal_shape(), style.shape, delay: 0.24)
            fact(L10n.reveal_punctuation(), style.punctuation, delay: 0.36)
        }
        .padding(.top, 4)
    }

    /// "Usually about 70 words", the number drawn large in the sentence the catalog words.
    private func length(_ count: Int) -> Text {
        var sentence = AttributedString(L10n.reveal_length(count: count))
        sentence.foregroundColor = Color.secondary
        sentence.font = WizardFont.text
        if let number = sentence.range(of: String(count)) {
            sentence[number].font = .system(size: lengthSize, weight: .semibold, design: .rounded).monospacedDigit()
            sentence[number].foregroundColor = Color.primary
        }
        return Text(sentence)
    }

    @ViewBuilder
    private func fact(_ label: String, _ value: String, delay: Double) -> some View {
        if !value.isEmpty {
            VStack(alignment: .leading, spacing: 3) {
                Text(label).font(WizardFont.label).foregroundStyle(.secondary)
                Text(value)
                    .font(WizardFont.text)
                    .fixedSize(horizontal: false, vertical: true)
            }
            .accessibilityElement(children: .combine)
            .wizardEntrance(shown, .rise, delay: delay)
        }
    }
}

/// Step 3: greetings and sign-offs, each over a bar as long as its part of its list.
struct RevealHabitsPage: View {
    let style: LanguageStyleRow
    let underTopBar: Bool

    @State private var shown = false
    @Environment(\.colorScheme) private var scheme
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @ScaledMetric(relativeTo: .title3) private var signsAsSize: CGFloat = 20

    var body: some View {
        WizardPage(title: L10n.reveal_step_habits(), underTopBar: underTopBar) {
            VStack(alignment: .leading, spacing: 18) {
                group(L10n.reveal_greetings(), style.greetings, start: 0.08)
                group(L10n.reveal_sign_offs(), style.signOffs, start: 0.32)
                if !style.signsAs.isEmpty {
                    HStack(alignment: .firstTextBaseline, spacing: 12) {
                        Text(L10n.reveal_signs_as()).font(WizardFont.label).foregroundStyle(.secondary)
                        Text(style.signsAs).font(.system(size: signsAsSize, weight: .medium, design: .serif))
                    }
                    .padding(.top, 2)
                    .accessibilityElement(children: .combine)
                    .wizardEntrance(shown, .rise, delay: 0.56)
                }
            }
        }
        .onAppear { shown = true }
    }

    @ViewBuilder
    private func group(_ label: String, _ habits: [HabitRow], start: Double) -> some View {
        if !habits.isEmpty {
            VStack(alignment: .leading, spacing: 6) {
                Text(label)
                    .font(WizardFont.label)
                    .foregroundStyle(.secondary)
                    .accessibilityAddTraits(.isHeader)
                VStack(spacing: 5) {
                    ForEach(Array(habits.enumerated()), id: \.offset) { index, habit in
                        row(habit, top: index == 0, delay: start + 0.08 * Double(index))
                    }
                }
            }
        }
    }

    private func row(_ habit: HabitRow, top: Bool, delay: Double) -> some View {
        let palette = WizardPalette(scheme)
        return HStack(spacing: 12) {
            RevealRunsText(runs: revealRuns(habit.text), font: WizardFont.habit)
            Spacer(minLength: 8)
            Text(revealFrequencyText(habit.frequency))
                .font(WizardFont.small)
                .foregroundStyle(.secondary)
                .fixedSize()
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 4)
        .frame(minHeight: 34)
        .background {
            ZStack(alignment: .leading) {
                RoundedRectangle(cornerRadius: 9).fill(palette.track)
                GrowingBar(fraction: shown || reduceMotion ? Double(habit.relative) / 100 : 0, cornerRadius: 9)
                    .fill(top ? palette.accentSoft : palette.fillStrong)
                    .animation(reduceMotion ? nil : WizardMotion.grow(delay), value: shown)
            }
        }
        .accessibilityElement(children: .combine)
    }
}

/// A greeting or sign-off, a placeholder such as `[Name]` drawn as a small pill in the accent.
struct RevealRunsText: View {
    let runs: [RevealRun]
    let font: Font

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        if runs.allSatisfy({ !$0.placeholder }) {
            Text(runs.map(\.text).joined()).font(font)
        } else {
            HStack(alignment: .firstTextBaseline, spacing: 0) {
                ForEach(Array(runs.enumerated()), id: \.offset) { _, run in
                    if run.placeholder {
                        pill(run.text)
                    } else {
                        Text(run.text).font(font)
                    }
                }
            }
        }
    }

    private func pill(_ text: String) -> some View {
        Text(text)
            .font(WizardFont.pill)
            .foregroundStyle(Color.accentColor)
            .padding(.horizontal, 7)
            .frame(minHeight: 18)
            .background(WizardPalette(scheme).accentFaint, in: Capsule())
            .overlay(Capsule().strokeBorder(Color.accentColor.opacity(0.4), lineWidth: 0.5))
            .padding(.horizontal, 1)
    }
}
